lib/coffee_place/result.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module CoffeePlace
+  Result = Struct.new(:success?, :value, :error) do
+    def failure?
+      !success?
+    end
+
+    class << self
+      def success(value)
+        new(true, value, nil)
+      end
+
+      def failure(error)
+        new(false, nil, error)
+      end
```

Comment:
Interesting approach - what concept are you using here to describe happy/unhappy paths?

Reply:
I'm using a sum type, with two variants.
This is similar to `Result` in Rust, or the `Maybe` type from functional languages.

The happy path should always hold a helpful value.
The unhappy paths always return an error message.

A better approach might be to have the `failure` cases return any
object that responds to a `message` method - allowing for various
ducked type error message objects to be passed in.

I think that using string error messages for the failure cases is sufficient
for a small example app like this.

Reply:
So, basically you're describing [monads](https://dry-rb.org/gems/dry-monads/1.3/result/) :stuck_out_tongue:

Reply:
Poor man's monads, without `fmap` and `chain/bind`.
But yes! :D

Reply:
I've used it in a few projects, it worked quite great, especially when used across all the layers of the project, as failures could be wrapped in a `Failure` either at the repository level or business layer and it would be passed along all the way to the presentation one where it would get handled appropriately, basically making it possible to have _no need for rescues_, awesome stuff.

You mentioned above that you enjoy operation chaining as in Elixir - `dry-monads` can help in this quite nicely, something like:

```ruby
def find_or_create(params)
Try { some_repo.find(params[:id]) } # raise an exception since the record doesn't exist
  .fmap(&:itself)
  .fmap { |record| <process_record> }
  .or { some_repo.create(params.except(:id)) }
```